### PR TITLE
clone settings corporate credit card export settings

### PR DESCRIPTION
**Pull request by NileshPant1999**

### CODE REVIEW

File: src/app/integration/onboarding/clone-settings/clone-settings.component.html [hunk 1]
```diff
@@ -199,13 +198,166 @@ <h5 class="clone-settings--header-caption sub-text-color">In this section, you c
...
+
+                    <div class="clone-settings--configuration-section">
+                        <div fxLayout="row">
+                            <div class="clone-settings--field-section" fxLayout="row" fxLayoutAlign="start center">
+                                <img class="clone-settings--sub-options-icon"
+                                    src="assets/images/svgs/general/document-disabled.svg">
+                                <p class="clone-settings--sub-options-text">Mode of Export</p>
+                                <img class="clone-settings--info-icon"
+                                    src="assets/images/svgs/general/info-disabled.svg"
+                                    matTooltip="Reimbursable expense in Fyle will be exported to Quickbooks"
```
Comment (ashwin1111) [inline src/app/integration/onboarding/clone-settings/clone-settings.component.html:254]: Credit Card expense

File: src/app/integration/onboarding/clone-settings/clone-settings.component.html [hunk 2]
```diff
@@ -199,13 +198,166 @@ <h5 class="clone-settings--header-caption sub-text-color">In this section, you c
...
+                                    matTooltipPosition="above" matTooltipClass="clone-settings">
+                            </div>
+                            <app-select [form]="cloneSettingsForm" [options]="cccExpenseExportOptions"
+                                [isFieldMandatory]="true" [mandatoryErrorListName]="'Purchase Bill Date'"
```
Comment (ashwin1111) [inline src/app/integration/onboarding/clone-settings/clone-settings.component.html:258]: mandatoryErrorListName is wrong

File: src/app/integration/onboarding/clone-settings/clone-settings.component.html [hunk 3]
```diff
@@ -199,13 +198,166 @@ <h5 class="clone-settings--header-caption sub-text-color">In this section, you c
...
+
+                    <div class="clone-settings--configuration-section">
+                        <div fxLayout="row">
+                            <div class="clone-settings--field-section" fxLayout="row" fxLayoutAlign="start center">
+                                <img class="clone-settings--sub-options-icon"
+                                    src="assets/images/svgs/general/tabs-disabled.svg">
+                                <p class="clone-settings--sub-options-text">How should the expenses be grouped?</p>
+                                <img class="clone-settings--info-icon"
+                                    src="assets/images/svgs/general/info-disabled.svg"
+                                    matTooltip="You could choose to export expenses when they have been approved and are awaiting payment clearance, OR simply when they have been paid out."
```
Comment (ashwin1111) [inline src/app/integration/onboarding/clone-settings/clone-settings.component.html:273]: tooltip is incorrect

File: src/app/integration/onboarding/clone-settings/clone-settings.component.html [hunk 4]
```diff
@@ -199,13 +198,166 @@ <h5 class="clone-settings--header-caption sub-text-color">In this section, you c
...
+
+                    <div class="clone-settings--configuration-section">
+                        <div fxLayout="row">
+                            <div class="clone-settings--field-section" fxLayout="row" fxLayoutAlign="start center">
+                                <img class="clone-settings--sub-options-icon"
+                                    src="assets/images/svgs/general/calendar-disabled.svg">
+                                <p class="clone-settings--sub-options-text">Date of export</p>
+                                <img class="clone-settings--info-icon"
+                                    src="assets/images/svgs/general/info-disabled.svg"
+                                    matTooltip="The selected date will reflect in expenses exported to Quickbooks Online"
```
Comment (ashwin1111) [inline src/app/integration/onboarding/clone-settings/clone-settings.component.html:292]: QuickBooks Online everywhere. B caps

File: src/app/integration/onboarding/clone-settings/clone-settings.component.html [hunk 5]
```diff
@@ -199,13 +198,166 @@ <h5 class="clone-settings--header-caption sub-text-color">In this section, you c
...
+                                    matTooltipPosition="above" matTooltipClass="clone-settings">
+                            </div>
+
+                            <app-select [form]="cloneSettingsForm" [options]="cccExpenseGroupingDateOptions"
+                                [isFieldMandatory]="true" [placeholder]="'Select the Purchase Bill date'"
```
Comment (ashwin1111) [inline src/app/integration/onboarding/clone-settings/clone-settings.component.html:297]: placeholder should be dynamic

File: src/app/integration/onboarding/clone-settings/clone-settings.component.html [hunk 6]
```diff
@@ -199,13 +198,166 @@ <h5 class="clone-settings--header-caption sub-text-color">In this section, you c
...
+                                    src="assets/images/svgs/general/document-disabled.svg">
+                                <p class="clone-settings--sub-options-text">Mode of Export</p>
+                                <img class="clone-settings--info-icon"
+                                    src="assets/images/svgs/general/info-disabled.svg"
+                                    matTooltip="Reimbursable expense in Fyle will be exported to Quickbooks"
+                                    matTooltipPosition="above" matTooltipClass="clone-settings">
+                            </div>
+                            <app-select [form]="cloneSettingsForm" [options]="cccExpenseExportOptions"
+                                [isFieldMandatory]="true" [mandatoryErrorListName]="'Purchase Bill Date'"
+                                [placeholder]="'Select the Purchase Bill date'"
```
Comment (ashwin1111) [inline src/app/integration/onboarding/clone-settings/clone-settings.component.html:259]: placeholder should be dynamic

File: src/app/integration/onboarding/clone-settings/clone-settings.component.html [hunk 7]
```diff
@@ -199,13 +198,166 @@ <h5 class="clone-settings--header-caption sub-text-color">In this section, you c
...
+
+                    <div class="clone-settings--configuration-section">
+                        <div fxLayout="row">
+                            <div class="clone-settings--field-section" fxLayout="row" fxLayoutAlign="start center">
+                                <img class="clone-settings--sub-options-icon"
+                                    src="assets/images/svgs/general/calendar-disabled.svg">
+                                <p class="clone-settings--sub-options-text">Set Default Credit Card Account as</p>
+                                <img class="clone-settings--info-icon"
+                                    src="assets/images/svgs/general/info-disabled.svg"
+                                    matTooltip="The selected date will reflect in expenses exported to Quickbooks Online"
```
Comment (ashwin1111) [inline src/app/integration/onboarding/clone-settings/clone-settings.component.html:311]: tooltip is incorrect

File: src/app/integration/onboarding/clone-settings/clone-settings.component.ts [hunk 8]
```diff
@@ -112,20 +114,46 @@ export class CloneSettingsComponent implements OnInit {
 
   getExportType(exportType: ReimbursableExpensesObject): string {
     const lowerCaseWord = exportType.toLowerCase();
-
     return lowerCaseWord.charAt(0).toUpperCase() + lowerCaseWord.slice(1);
   }
 
+  setCreditCardExpenseGroupingDateOptions(creditCardExportGroup: ExpenseGroupingFieldOption): void {
```
Comment (ashwin1111) [inline src/app/integration/onboarding/clone-settings/clone-settings.component.ts:120]: private

File: src/app/integration/onboarding/clone-settings/clone-settings.component.ts [hunk 9]
```diff
@@ -112,20 +114,46 @@ export class CloneSettingsComponent implements OnInit {
...
+    }
+  }
+
   showExpenseAccountField(): boolean {
     return this.cloneSettingsForm.controls.reimbursableExportType.value === ReimbursableExpensesObject.EXPENSE;
   }
 
   showBankAccountField(): boolean {
-    return this.employeeFieldMapping === EmployeeFieldMapping.EMPLOYEE && this.cloneSettingsForm.controls.reimbursableExportType.value && this.cloneSettingsForm.controls.reimbursableExportType.value !== ReimbursableExpensesObject.EXPENSE;
+    return this.cloneSettingsForm.controls.employeeMapping.value === EmployeeFieldMapping.EMPLOYEE && this.cloneSettingsForm.controls.reimbursableExportType.value && this.cloneSettingsForm.controls.reimbursableExportType.value !== ReimbursableExpensesObject.EXPENSE;
```
Comment (ashwin1111) [inline src/app/integration/onboarding/clone-settings/clone-settings.component.ts:136]: use `this.cloneSettingsForm.value.employeeMapping`

File: src/app/integration/onboarding/clone-settings/clone-settings.component.ts [hunk 10]
```diff
@@ -112,20 +114,46 @@ export class CloneSettingsComponent implements OnInit {
...
   }
 
   showReimbursableAccountsPayableField(): boolean {
-    return (this.cloneSettingsForm.controls.reimbursableExportType.value === ReimbursableExpensesObject.BILL) || (this.cloneSettingsForm.controls.reimbursableExportType.value === ReimbursableExpensesObject.JOURNAL_ENTRY && this.employeeFieldMapping === EmployeeFieldMapping.VENDOR);
+    return (this.cloneSettingsForm.controls.reimbursableExportType.value === ReimbursableExpensesObject.BILL) || (this.cloneSettingsForm.controls.reimbursableExportType.value === ReimbursableExpensesObject.JOURNAL_ENTRY && this.cloneSettingsForm.controls.employeeMapping.value === EmployeeFieldMapping.VENDOR);
```
Comment (ashwin1111) [inline src/app/integration/onboarding/clone-settings/clone-settings.component.ts:140]: same as above

File: src/app/integration/onboarding/clone-settings/clone-settings.component.ts [hunk 11]
```diff
@@ -136,44 +164,169 @@ export class CloneSettingsComponent implements OnInit {
...
+      this.cloneSettingsForm.controls.accountsPayable.clearValidators();
+      this.cloneSettingsForm.controls.accountsPayable.updateValueAndValidity();
+    }
+
+    if (this.showDefaultCreditCardVendorField()) {
+      this.cloneSettingsForm.controls.defaultCreditCardVendor.setValidators(Validators.required);
+    } else {
+      this.cloneSettingsForm.controls.defaultCreditCardVendor.clearValidators();
+      this.cloneSettingsForm.controls.defaultCreditCardVendor.updateValueAndValidity();
+    }
```
Comment (ashwin1111) [inline src/app/integration/onboarding/clone-settings/clone-settings.component.ts:194]: can we move these to services if these are common for both components?

File: src/app/integration/onboarding/clone-settings/clone-settings.component.ts [hunk 12]
```diff
@@ -136,44 +164,169 @@ export class CloneSettingsComponent implements OnInit {
...
+          label: 'Spend Date',
+          value: ExportDateType.SPENT_AT
+        }];
     } else {
-      this.cloneSettingsForm.controls.accountsPayable.clearValidators();
-      this.cloneSettingsForm.controls.accountsPayable.updateValueAndValidity();
+      this.cloneSettingsForm.controls.creditCardExportGroup.enable();
+      this.setCreditCardExpenseGroupingDateOptions(this.cloneSettingsForm.controls.creditCardExportGroup.value);
     }
   }
```
Comment (ashwin1111) [inline src/app/integration/onboarding/clone-settings/clone-settings.component.ts:221]: can we move these to services if these are common for both components?

File: src/app/integration/onboarding/clone-settings/clone-settings.component.ts [hunk 13]
```diff
@@ -136,44 +164,169 @@ export class CloneSettingsComponent implements OnInit {
...
     this.setGeneralMappingsValidator();
   }
 
   private setupForm(): void {
     this.cloneSettingsForm = this.formBuilder.group({
       // Employee Mapping
-      employeeMapping: [this.cloneSettings.employee_mappings.workspace_general_settings.employee_field_mapping],
-      autoMapEmployee: [this.cloneSettings.employee_mappings.workspace_general_settings.auto_map_employees],
+      employeeMapping: [this.cloneSettings.employee_mappings.workspace_general_settings?.employee_field_mapping, Validators.required],
+      autoMapEmployee: [this.cloneSettings.employee_mappings.workspace_general_settings?.auto_map_employees, Validators.nullValidator],
```
Comment (ashwin1111) [inline src/app/integration/onboarding/clone-settings/clone-settings.component.ts:309]: why are we adding `Validators.nullValidator`?

Comment (NileshPant1999) [reply src/app/integration/onboarding/clone-settings/clone-settings.component.ts:309]: we were doing this employee setting form

```
        employeeMapping: [this.existingEmployeeFieldMapping, Validators.required],
        autoMapEmployee: [responses[0].workspace_general_settings?.auto_map_employees, Validators.nullValidator]
```

Comment (ashwin1111) [reply src/app/integration/onboarding/clone-settings/clone-settings.component.ts:309]: doesn't add any value, let's remove it off

File: src/app/integration/onboarding/clone-settings/clone-settings.component.ts (hunk 12 above)
Comment (NileshPant1999) [reply src/app/integration/onboarding/clone-settings/clone-settings.component.ts:178]: @ashwin1111 i tried, but it needs a lot of change then in both clone settings and export settings form i will try that again in the final pr.

Comment (ashwin1111) [reply src/app/integration/onboarding/clone-settings/clone-settings.component.ts:178]: sure can skip if it's getting complex

**Review by ashwin1111 — APPROVED** [review]

**Pull request merged**